src/attachment.rs

```diff
@@ -0,0 +1,344 @@
+    type Target = UnsafeCell<Option<HashMap<Cow<'static, [u8]>, Cow<'static, [u8]>>>>;
+    fn deref(&self) -> &Self::Target {
+        unsafe { core::mem::transmute(self) }
+    }
+}
+
+/// Constructs a new map.
+#[no_mangle]
+pub extern "C" fn z_bytes_map_new() -> z_owned_bytes_map_t {
+    unsafe { core::mem::transmute(Some(HashMap::<Cow<[u8]>, Cow<[u8]>>::new())) }
+}
+
+/// Constructs the gravestone value for `z_owned_bytes_map_t`
+#[no_mangle]
+pub extern "C" fn z_bytes_map_null() -> z_owned_bytes_map_t {
+    unsafe { core::mem::transmute(None::<HashMap<Cow<[u8]>, Cow<[u8]>>>) }
+}
+
+/// Returns `true` if the map is not in its gravestone state
+#[no_mangle]
+pub extern "C" fn z_bytes_map_check(this: &z_owned_bytes_map_t) -> bool {
+    unsafe { &*this.get() }.is_some()
+}
+/// Destroys the map, resetting `this` to its gravestone value.
+///
+/// This function is double-free safe, passing a pointer to the gravestone value will have no effect.
+#[no_mangle]
+pub extern "C" fn z_bytes_map_drop(this: &mut z_owned_bytes_map_t) {
+    let this = unsafe { &mut *this.get() };
+    this.take();
+}
+
+/// Returns the value associated with `key`, returning a gravestone value if:
+/// - `this` or `key` is in gravestone state.
+/// - `this` has no value associated to `key`
+#[no_mangle]
+pub extern "C" fn z_bytes_map_get(this: &z_owned_bytes_map_t, key: z_bytes_t) -> z_bytes_t {
+    let this = unsafe { &*this.get() };
+    let (Some(this), Some(key)) = (this.as_ref(), key.as_slice()) else {
+        return z_bytes_null();
+    };
+    if let Some(value) = this.get(key) {
+        value.as_ref().into()
+    } else {
+        z_bytes_null()
+    }
+}
+
+/// Associates `value` to `key` in the map, copying them to obtain ownership: `key` and `value` are not aliased past the function's return.
+///
+/// Calling this with `NULL` or the gravestone value is undefined behaviour.
+#[no_mangle]
+pub extern "C" fn z_bytes_map_insert_by_copy(
+    this: &z_owned_bytes_map_t,
+    key: z_bytes_t,
+    value: z_bytes_t,
+) {
+    let this = unsafe { &mut *this.get() };
+    if let (Some(this), Some(key), Some(value)) = (this.as_mut(), key.as_slice(), value.as_slice())
+    {
+        this.insert(Cow::Owned(key.to_owned()), Cow::Owned(value.to_owned()));
+    }
+}
+
+/// Associates `value` to `key` in the map, aliasing them.
+///
+/// Note that once `key` is aliased, reinserting at the same key may alias the previous instance, or the new instance of `key`.
+///
+/// Calling this with `NULL` or the gravestone value is undefined behaviour.
+#[no_mangle]
+pub extern "C" fn z_bytes_map_insert_by_alias(
+    this: &z_owned_bytes_map_t,
+    key: z_bytes_t,
+    value: z_bytes_t,
+) {
+    let this = unsafe { &mut *this.get() };
+    if let (Some(this), Some(key), Some(value)) = (this.as_mut(), key.as_slice(), value.as_slice())
+    {
+        unsafe {
+            this.insert(
+                Cow::Borrowed(core::mem::transmute(key)),
```

Comment:
Done

Reply:
I strongly oppose this: by making every `as_slice` return `'static` slices, we're losing a safety net against accidentally introducing a dependency on `z_bytes_t` outliving the function it's passed to.

We're taking the risk of accidentally aliasing parameters without noticing, all to avoid the "safest" usage of a `transmute` in a function that's explicitly documented to have a constraint on the parameter's lifetime.

Sorry @sashacmc, but can you revert this change?

Reply:
Done!